torch/_refs/__init__.py

```diff
@@ -1655,48 +1655,108 @@ def item(a: TensorLikeType) -> NumberType:
+            or utils.is_contiguous_for_memory_format(a, memory_format=memory_format)
         )
     )
 
 
-# TODO: there are multiple overloads of to, but we only support one here
-# aten/src/ATen/native/native_functions.yaml lists all overloads
-# Missing overloads:
-# to.dtype_layout
-# to.device
-# to.other
-# https://github.com/pytorch/pytorch/issues/84264
+def _to_dispatch(
+    device=None,
+    dtype=None,
+    non_blocking=None,
+    copy=None,
+    memory_format=None,
+    layout=None,
+    pin_memory=None,
+):
+    if isinstance(device, TensorLike):
+        # overload to `to.other`
+        if isinstance(copy, torch.memory_format):
+            memory_format = copy
+        # we need to make sure that dtype is bool, otherwise, non_blocking could be passed as kwarg
+        if isinstance(non_blocking, bool) and isinstance(dtype, bool):
+            copy = non_blocking
+        if isinstance(dtype, bool):
+            non_blocking = dtype
+        # extract tensor options
+        other = device
+        device = other.device
+        dtype = other.dtype
+        layout = other.layout
+        pin_memory = other.pin_memory
+    elif isinstance(device, torch.dtype):
+        # overload to `to.dtype`
+        if isinstance(copy, torch.memory_format):
+            memory_format = copy
+        # we need to make sure that dtype is bool, otherwise, non_blocking could be passed as kwarg
+        if isinstance(non_blocking, bool) and isinstance(dtype, bool):
+            copy = non_blocking
+        if isinstance(dtype, bool):
+            non_blocking = dtype
+        dtype = device
+        layout = None
+        pin_memory = None
+        device = None
+    # overload to `to.device`
+    # do nothing because signature of 'to' matches
+    # overload to `to.dtype_layout`
+    # do nothing because the `to.dtype_layout` takes all kwargs.
+    return device, dtype, non_blocking, copy, memory_format, layout, pin_memory
+
+
 def to(
     a: TensorLikeType,
-    dtype: torch.dtype,
-    non_blocking: bool = False,
-    copy: bool = False,
-    memory_format: torch.memory_format = torch.preserve_format,
+    device: Optional[torch.device] = None,
+    dtype: Optional[torch.dtype] = None,
+    non_blocking: Optional[bool] = None,
+    copy: Optional[bool] = None,
+    memory_format: Optional[torch.memory_format] = None,
+    *,
+    layout: Optional[torch.layout] = None,
+    pin_memory: Optional[bool] = None,
 ) -> TensorLikeType:
-    if _to_will_alias(a, dtype, copy, memory_format):
+    device, dtype, non_blocking, copy, memory_format, layout, pin_memory = _to_dispatch(
+        device, dtype, non_blocking, copy, memory_format, layout, pin_memory
+    )
+
+    if _to_will_alias(a, device, dtype, copy, memory_format, layout):
         return a
     if (
-        (copy is True or dtype != a.dtype)
-        and memory_format == torch.preserve_format
-        and non_blocking is False
+        (copy or dtype != a.dtype)
```

Comment:
Let's add a comment here that this is a short-circuit for when just a dtype change is requested.

Seems like this short-circuit could be more aggressive?

Reply:
Will update the short-circuit condition~